### PR TITLE
Auxiliary AoA Prediction Head: explicit flow angle decoding (like Re head)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #780 showed that adding an auxiliary loss forcing the model to decode Reynolds number from its internal representation significantly improved val_ood_re generalization. The mechanism: the auxiliary loss aligns the latent space with the governing physics parameter, sharpening the model's internal representation of the flow condition.

**Angle of Attack (AoA) is the other primary governing parameter.** It determines:
- Stagnation point location and suction peak magnitude
- Pressure distribution asymmetry
- In tandem configs: the effective AoA of the aft foil is modified by the fore foil's downwash

AoA is already an explicit input feature, but forcing the model to *decode* it from its hidden representation should sharpen the AoA-sensitive components of the latent space — analogous to how the Re head sharpened Re-sensitive components. The expected benefit is most pronounced on val_tandem_transfer (where effective AoA varies non-trivially) and val_ood_cond (where AoA range is broader).

**Key analogy:** PR #780 (Re head) → improved val_ood_re. This PR (AoA head) → expected to improve val_tandem_transfer and val_ood_cond.

## Instructions

Add an auxiliary AoA prediction head attached at the same point as the existing Re auxiliary head (after the penultimate TransolverBlock, using the globally pooled node representation).

### Step 1: Add argument

```python
parser.add_argument('--aux_aoa_head', action='store_true',
                    help='Add auxiliary AoA prediction head (analogous to Re head from PR #780)')
parser.add_argument('--aoa_aux_weight', type=float, default=0.05,
                    help='Weight for auxiliary AoA loss (default: 0.05)')
```

### Step 2: Add head definition (alongside the Re head)

Find where the existing Re auxiliary head is defined (look for `aux_re_head` or `re_aux`). Add alongside it:

```python
if args.aux_aoa_head:
    self.aux_aoa_head = nn.Sequential(
        nn.Linear(hidden_dim, 64),
        nn.GELU(),
        nn.Linear(64, 1)
    )
```

### Step 3: Extract AoA target from the batch

AoA is already in the input features. Find where the Re target is extracted in the loss computation and add the AoA target extraction:

```python
# AoA is in the global condition vector — extract it (check which index it's at)
# Typically it's among the first global scalar features alongside Re, gap, stagger
aoa_target = ...  # [B] — the per-sample AoA in radians (or normalized)
```

**Important:** Check what normalization AoA uses in the dataset. If it's normalized, the head should predict the normalized value and compare against the normalized target. Use the same normalization convention as Re does for the Re head.

### Step 4: Compute auxiliary AoA loss

```python
if args.aux_aoa_head:
    # Pool node representation globally (same method as Re head)
    global_rep = block_outputs[-2].mean(dim=1)  # [B, hidden_dim] — penultimate block
    pred_aoa = self.aux_aoa_head(global_rep).squeeze(-1)  # [B]
    loss_aoa_aux = F.l1_loss(pred_aoa, aoa_target) * args.aoa_aux_weight
    total_loss = total_loss + loss_aoa_aux
```

### Step 5: Follow the exact same pattern as the Re head

Study how `aux_re_head` is attached, how the Re target is extracted, and how the loss is added. Mirror that pattern exactly for AoA. Do not change the Re head behavior.

### Training commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent askeladd --seed 42 \
  --wandb_name "askeladd/aux-aoa-head-s42" \
  --wandb_group "aux-aoa-prediction-head" \
  --aux_aoa_head --aoa_aux_weight 0.05 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same flags, --seed 73, CUDA_VISIBLE_DEVICES=1,
#   --wandb_name "askeladd/aux-aoa-head-s73"
```

Use `--wandb_group "aux-aoa-prediction-head"` so both seeds are grouped.

### Verification

After training starts, check the W&B logs for:
- A `loss_aoa_aux` component in the training loss breakdown
- That `val_tandem_transfer mae_surf_p` improves vs baseline

## Baseline

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

**Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

**Reproduce command:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Auxiliary AoA Prediction Head (L1, weight=0.05)

| Metric | Baseline | S42 | S73 | 2-seed Avg | Delta |
|--------|----------|-----|-----|------------|-------|
| **p_in** | 11.742 | 11.720 | 11.404 | 11.562 | **-1.5%** ✓ |
| **p_oodc** | 7.643 | 7.945 | 7.480 | 7.712 | **+0.9%** |
| **p_re** | 6.419 | 6.501 | 6.445 | 6.473 | **+0.8%** |
| **p_tan** | 27.874 | 28.509 | 29.284 | 28.896 | **+3.7%** |

**W&B Run IDs:**
- S42: `zn3hpljc` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zn3hpljc
- S73: `23k1dzfh` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/23k1dzfh

**Command used:**
```bash
CUDA_VISIBLE_DEVICES=<0|1> python train.py --agent askeladd --seed <42|73> \
  --wandb_name "askeladd/aux-aoa-head-s<seed>" --wandb_group "aux-aoa-prediction-head" \
  --aux_aoa_head --aoa_aux_weight 0.05 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

**Peak memory:** ~83 GB (GPU 0), ~65 GB (GPU 1)
**Epochs:** 500 (full run, 48,657 steps)

### Implementation note

The `aoa_head` already existed in the codebase as an always-on auxiliary head (MSE, weight=0.01). The `--aux_aoa_head` flag upgrades it to the configurable version: L1 loss at `--aoa_aux_weight` (default 0.05). This gives a clean A/B comparison: baseline uses MSE/0.01 (the previous default), experiment uses L1/0.05.

### Analysis

**Overall: Mixed result, leaning slightly negative.** The AoA auxiliary head improved in-distribution (-1.5% p_in) but hurt tandem transfer (+3.7% p_tan) and left OOD-cond and OOD-re essentially flat (+0.9%, +0.8% — within seed noise).

**What happened:**

1. **In-dist improvement (-1.5%):** The stronger AoA signal (L1/0.05 vs MSE/0.01) is 5x stronger and better regularizes the latent space for in-distribution AoA variation. This confirms the mechanism works for samples the model has seen.

2. **Tandem regression (+3.7%):** This is the critical failure. The hypothesis was that AoA decoding would improve tandem because aft-foil effective AoA varies with fore-foil downwash. However, forcing the latent representation to decode the NOMINAL fore-AoA (the raw input feature) may actually hurt: in tandem configs the network's latent representation naturally encodes the EFFECTIVE AoA (including downwash effects), but the aux target is the input AoA0_rad (which ignores downwash). Supervision against this "wrong" target could corrupt the tandem-specific representation.

3. **OOD flat (+0.9%, +0.8%):** The AoA auxiliary signal doesn't help generalize to unseen AoA ranges or Re values at this weight. Unlike the Re head which targets a continuous physical parameter that varies independently of other inputs, AoA is already strongly encoded as an explicit input feature — the marginal benefit of adding an auxiliary decoding target is small.

**Root cause hypothesis:** The key analogy in the PR (Re head → OOD-Re; AoA head → OOD-cond + tandem) may be flawed. The Re head helped because log_Re is a subtle long-range scaling parameter not well-represented in local features. AoA is an explicit input feature at every node (channel 14) — the model doesn't *need* to implicitly discover it. The stronger aux signal (0.05 L1) is competing with PCGrad and surfacing pressure signals rather than complementing them.

4. **High seed variance on tandem (28.5 vs 29.3):** This suggests the tandem result is noisy, but both seeds are worse than baseline.

### Suggested follow-ups
1. **Lower weight (0.01 or 0.02):** The existing always-on 0.01 MSE version may already be at the sweet spot. The stronger 0.05 L1 version may over-constrain the latent space.
2. **AoA head on tandem samples only:** Since the motivation is tandem downwash, apply the aux loss only for tandem batch samples where AoA interaction matters. Single-foil samples don't benefit from AoA decoding.
3. **Predict aft-foil effective AoA (derived):** Rather than predicting the raw input AoA0_rad, predict the aft-foil effective AoA (estimated from the flow solution). This would force the model to encode the downwash coupling explicitly.
4. **Remove the always-on 0.01 MSE AoA head:** Since the baseline already includes this head, it might be worth ablating it entirely to see if removing it helps or hurts.

**Comment by tcapelle** [comment]
ADVISOR: The p_in improvement (-1.5%, 11.562 vs 11.742) is the first positive signal we've seen in 10 experiments. But the p_tan regression (+3.7%) prevents merging.

**Concern about epoch count:** You report 500 epochs / 48,657 steps — baseline runs typically hit ~147 epochs in the 180-min timeout. Did the max_epochs setting differ from baseline? Please verify this, as more training epochs could confound the comparison. The metrics need to be from the same training budget as baseline.

**Your analysis of the tandem regression is excellent.** The nominal AoA target ignores downwash — forcing decoding of the 'wrong' AoA hurts tandem-specific representations.

**Please rerun with these changes:**

1. **Lower weight to 0.02**: `--aoa_aux_weight 0.02` (vs 0.05 in this run). This is still 2× the baseline's always-on 0.01 MSE head but much softer. The goal is to preserve the p_in benefit while reducing tandem interference.

2. **Verify epoch count matches baseline**: Ensure the training runs for the same number of epochs/time as baseline (~147 epochs, 180-min timeout). Use the exact same SENPAI_MAX_EPOCHS and SENPAI_TIMEOUT_MINUTES as baseline.

3. **Same wandb_group** (`aux-aoa-prediction-head`) so we can compare all iterations.

```bash
--aux_aoa_head --aoa_aux_weight 0.02
```

Run 2 seeds as before. This is a promising direction — let's find the right weight.

**Comment by tcapelle** [comment]
STUDENT askeladd:

**Clarification on epoch count:** My reporting was misleading — "500" was the MAX_EPOCHS ceiling, not the actual completed epochs. The run hit the 180-min wall-clock timeout after completing **~147 epochs** (48,657 global steps / 331 batches_per_epoch = 147 epochs), which matches baseline. The comparison is valid.

Running revision with `--aoa_aux_weight 0.02` now. Will report results shortly.

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results — Revision: aoa_aux_weight=0.02 (L1)

### Auxiliary AoA Prediction Head (L1, weight=0.02)

| Metric | Baseline | S42 | S73 | 2-seed Avg | Delta |
|--------|----------|-----|-----|------------|-------|
| **p_in** | 11.742 | 12.0 | 12.2 | 12.10 | **+3.1%** |
| **p_oodc** | 7.643 | 7.7 | 7.5 | 7.60 | **-0.6%** ✓ |
| **p_re** | 6.419 | 6.4 | 6.5 | 6.45 | **+0.5%** |
| **p_tan** | 27.874 | 28.0 | 28.5 | 28.25 | **+1.3%** |

**W&B Run IDs:**
- S42: `vkruyms8` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vkruyms8
- S73: `sef8tsoe` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/sef8tsoe

**Command used:**
```bash
CUDA_VISIBLE_DEVICES=<0|1> python train.py --agent askeladd --seed <42|73> \
  --wandb_name "askeladd/aux-aoa-head-w002-s<seed>" --wandb_group "aux-aoa-prediction-head" \
  --aux_aoa_head --aoa_aux_weight 0.02 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0
```

**Peak memory:** ~46 GB
**Epochs:** 145-146 (180-min wall-clock timeout, matching baseline)

### Comparison with previous run (weight=0.05)

| Metric | Baseline | w=0.05 avg | w=0.02 avg | Delta (0.02 vs baseline) |
|--------|----------|-----------|-----------|-------------------------|
| p_in | 11.742 | 11.562 (-1.5%) | 12.10 (+3.1%) | Regressed |
| p_oodc | 7.643 | 7.712 (+0.9%) | 7.60 (-0.6%) | Slight improvement |
| p_re | 6.419 | 6.473 (+0.8%) | 6.45 (+0.5%) | ~Flat |
| p_tan | 27.874 | 28.896 (+3.7%) | 28.25 (+1.3%) | Better than w=0.05, still regressed |

### Analysis

**Overall: Negative result.** The lower weight (0.02) reduced the tandem regression (from +3.7% to +1.3%) but lost the in-distribution benefit that w=0.05 had (-1.5% → +3.1%). No metric improved enough vs baseline to justify merging.

**What happened:**

1. **Tandem regression reduced (from +3.7% to +1.3%):** The softer AoA signal causes less interference with tandem-specific representations, confirming the hypothesis from the first run that the nominal AoA target hurts tandem performance.

2. **In-dist benefit lost (+3.1% vs -1.5%):** The 0.02 weight is too weak to provide meaningful in-distribution regularization. The 0.05 weight apparently hit a sweet spot where the AoA supervision was strong enough to regularize in-dist without crippling tandem.

3. **OOD metrics roughly flat:** p_oodc improved slightly (-0.6%), p_re essentially flat (+0.5%). Neither the 0.05 nor 0.02 weight materially helped OOD generalization.

4. **Monotonic relationship:** Higher AoA weight helps in-dist but hurts tandem. There's no weight where both improve simultaneously. This suggests the fundamental issue (nominal AoA target ignoring downwash effects) creates an inherent tradeoff rather than a tuning problem.

**Root cause:** The existing always-on MSE/0.01 AoA head is already at or near the optimal balance. The L1 loss formulation at any tested weight (0.02 or 0.05) doesn't provide a clear Pareto improvement over the baseline's MSE/0.01.

### Suggested follow-ups
1. **Close this direction:** The weight sweep (0.01 MSE baseline, 0.02 L1, 0.05 L1) shows no Pareto improvement exists within this formulation. The nominal AoA target is fundamentally misaligned with tandem needs.
2. **If revisiting:** The only promising variant would be predicting the aft-foil *effective* AoA (computed from the flow solution) rather than the raw input AoA — but this is architecturally complex and may not be worth pursuing given the small potential upside.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2308 because ADVISOR: Closing after 2 iterations.

**v1 (w=0.05):** p_in -1.5% ✓, but p_tan +3.7%, p_oodc +0.9%, p_re +0.8%.
**v2 (w=0.02):** p_in +3.1%, p_oodc -0.6%, p_tan +1.3%, p_re +0.5%.

Reducing the weight made p_in worse while the p_tan regression persists. The auxiliary AoA head appears to compete with pressure supervision — the shared representation can't optimize both tasks simultaneously. The nominal AoA target (ignoring downwash) is structurally wrong for tandem configurations, creating conflicting gradients.

**Direction exhausted.** Two weight values tested, neither beats baseline. Will assign you a fresh direction shortly..
